### PR TITLE
WIP: add the performance-addon-operator code

### PR DESCRIPTION
**Pull request by cynepco3hahue**
## TODO:
- [ ] e2e tests, not all of them but at least part of them.
- [ ] add owners file for the PAO component
- [ ] disabling the PAO operator based on a ManagementState
- [ ] reporting PAO operator status via co/node-tuning

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -51,7 +61,10 @@ replace (
 	k8s.io/kubernetes => k8s.io/kubernetes v1.21.1
 	k8s.io/legacy-cloud-providers => k8s.io/legacy-cloud-providers v0.21.1
 	k8s.io/metrics => k8s.io/metrics v0.21.1
+	k8s.io/mount-utils => k8s.io/mount-utils v0.21.2
 	k8s.io/sample-apiserver => k8s.io/sample-apiserver v0.21.1
+	sigs.k8s.io/controller-runtime => sigs.k8s.io/controller-runtime v0.9.2
```
Comment (jmencak) [inline go.mod:66]: Do we really need `sigs.k8s.io/controller-runtime`?

File: manifests/20-crd-profile.yaml [hunk 2]
```diff
@@ -16,102 +16,90 @@ spec:
...
-                  Tuned profile
-                type: string
-              conditions:
-                description: conditions represents the state of the per-node Profile
-                  application
-                type: array
-                items:
-                  description: ProfileStatusCondition represents a partial state of
-                    the per-node Profile application.
+    - additionalPrinterColumns:
```
Comment (jmencak) [inline manifests/20-crd-profile.yaml:19]: Why has this file changed?

File: Makefile [hunk 3]
```diff
@@ -48,21 +46,27 @@ clone-tuned:
 	  cd $(TUNED_DIR) && git checkout $(TUNED_COMMIT) && cd .. && \
 	  rm -rf $(TUNED_DIR)/.git)
 
-build: $(BINDATA) pkg/generated
+build: $(BINDATA) generate-deepcopy pkg/generated
```
Comment (jmencak) [inline Makefile:49]: Perhaps at some point add `performance-addon-operator-bin` target?

**Review by jmencak — COMMENTED** [review]
Thanks you for the PR, Artyom, it is a good start.  In order to pass the verification, you need to fix your formatting.  See "make verify" and it will tell you which files to fix.

I haven't gone through all of the code, just the interface and added a few comments.  PAO commpiled fine. I'd really like to get rid of the dependency on `sigs.k8s.io/controller-runtime` if possible.

**Comment by jmencak** [comment]
```
$ make verify
verify-gofmt: ERROR: gofmt failed on the following files:
./pkg/apis/performanceprofile/v2/register.go ./pkg/performance-addon-controller/components/manifestset/manifestset.go ./pkg/performance-addon-controller/components/tuned/tuned.go ./pkg/performance-addon-controller/performanceprofile_controller_suite_test.go ./pkg/performance-addon-controller/performanceprofile_controller_test.go ./pkg/performance-addon-controller/status.go

For details, run: gofmt -d -s ./pkg/apis/performanceprofile/v2/register.go ./pkg/performance-addon-controller/components/manifestset/manifestset.go ./pkg/performance-addon-controller/components/tuned/tuned.go ./pkg/performance-addon-controller/performanceprofile_controller_suite_test.go ./pkg/performance-addon-controller/performanceprofile_controller_test.go ./pkg/performance-addon-controller/status.go

make: *** [Makefile:116: verify-gofmt] Error 1

```

**Comment by jmencak** [comment]
Another thing we should consider is modifying the `OWNERS` file to add approvers from the PAO team.

**Comment by jmencak** [comment]
Another two issues to solve that currently come to mind:
  * disabling the PAO operator based on a [ManagementState](https://github.com/openshift/api/blob/67c28690af52a69e0b8fa565916fe1b9b7f52f10/operator/v1/types.go#L30)
  * reporting PAO operator status via [co/node-tuning](https://github.com/openshift/cluster-node-tuning-operator/blob/master/pkg/clusteroperator/status.go) ?

File: go.mod (hunk 1 above)
Comment (cynepco3hahue) [reply go.mod:66]: We are using it a lot under the PAO controller. In general, it makes it much easier to write a controller with it, without the need to specify separate clients for each API type, informers, and listers.

File: manifests/20-crd-profile.yaml (hunk 2 above)
Comment (cynepco3hahue) [reply manifests/20-crd-profile.yaml:19]: I just ran the `make crd-schema-gen`.

File: Makefile (hunk 3 above)
Comment (cynepco3hahue) [reply Makefile:49]: For sure, I will do it once will add PAO dockerfiles.

**Comment by cynepco3hahue** [comment]
> Thanks you for the PR, Artyom, it is a good start. In order to pass the verification, you need to fix your formatting. See "make verify" and it will tell you which files to fix.
> 
> I haven't gone through all of the code, just the interface and added a few comments. PAO commpiled fine. I'd really like to get rid of the dependency on `sigs.k8s.io/controller-runtime` if possible.

A lot of thanks for the initial review, we are using a lot `sigs.k8s.io/controller-runtime` under the PAO controller, it makes it much easier to configure the controller and I can see it pretty popular among OpenShift repositories.

File: go.mod (hunk 1 above)
Comment (jmencak) [reply go.mod:66]: It took me quite a bit of effort to get away from `operator-sdk` and `controller-runtime` in the past.  Perhaps things are better now, but it used to be quite hard (nearly impossible) to keep other k8s dependencies up-to-date when I was using those two dependencies in the past.  How easy/difficult is it now? I.e. use the current [1.22 k8s](https://github.com/openshift/cluster-node-tuning-operator/pull/261) deps?

**Comment by yanirq** [comment]
@cynepco3hahue  you can close this PR in favor of https://github.com/openshift/cluster-node-tuning-operator/pull/302  or keep it for additional references in the near future (maybe even as closed if we still have the code)

**Comment by cynepco3hahue** [comment]
close in favor of #302.

**Pull request closed without merging**